src/main/java/com/google/idea/perf/vfstracer/VirtualFileTracer.kt

```diff
@@ -0,0 +1,255 @@
+
+        val classVisitor = object: ClassVisitor(ASM_API, writer) {
+            override fun visitMethod(
+                access: Int,
+                name: String?,
+                descriptor: String?,
+                signature: String?,
+                exceptions: Array<out String>?
+            ): MethodVisitor {
+                if (name != "createPsiNoLock") {
+                    return super.visitMethod(access, name, descriptor, signature, exceptions)
+                }
+
+                val methodWriter = cv.visitMethod(access, name, descriptor, signature, exceptions)
+
+                return object: AdviceAdapter(ASM_API, methodWriter, access, name, descriptor) {
+                    override fun onMethodExit(opcode: Int) {
+                        mv.visitInsn(Opcodes.DUP)
+                        mv.visitMethodInsn(
+                            Opcodes.INVOKESTATIC,
+                            HOOK_CLASS_JVM_NAME,
+                            ON_PSI_ELEMENT_CREATE.name,
+                            ON_PSI_ELEMENT_CREATE.descriptor,
+                            false
+                        )
+                    }
+                }
+            }
+        }
+
+        reader.accept(classVisitor, SKIP_FRAMES)
+        return writer.toByteArray()
+    }
+
+    private fun tryTransformStubIndex(classBytes: ByteArray): ByteArray {
+        val reader = ClassReader(classBytes)
+        val writer = ClassWriter(reader, COMPUTE_MAXS or COMPUTE_FRAMES)
+
+        val classVisitor = object: ClassVisitor(ASM_API, writer) {
+            override fun visitMethod(
+                access: Int,
+                name: String?,
+                descriptor: String?,
+                signature: String?,
+                exceptions: Array<out String>?
+            ): MethodVisitor {
+                if (name != "processElements" || descriptor != "(Lcom/intellij/psi/stubs/StubIndexKey;Ljava/lang/Object;Lcom/intellij/openapi/project/Project;Lcom/intellij/psi/search/GlobalSearchScope;Lcom/intellij/util/indexing/IdFilter;Ljava/lang/Class;Lcom/intellij/util/Processor;)Z") {
+                    return super.visitMethod(access, name, descriptor, signature, exceptions)
+                }
+
+                val methodWriter = cv.visitMethod(access, name, descriptor, signature, exceptions)
+
+                return object: AdviceAdapter(ASM_API, methodWriter, access, name, descriptor) {
+                    override fun onMethodEnter() {
+                        mv.visitVarInsn(Opcodes.ALOAD, 7)
+                        mv.visitMethodInsn(
+                            Opcodes.INVOKESTATIC,
+                            HOOK_CLASS_JVM_NAME,
+                            ON_STUB_INDEX_PROCESSOR_CREATE.name,
+                            ON_STUB_INDEX_PROCESSOR_CREATE.descriptor,
+                            false
+                        )
+                        mv.visitVarInsn(Opcodes.ASTORE, 7)
```

Comment:
Nice work on figuring how to trace this by the way (swapping out the `processor` parameter) 😄

Reply:
Thanks!